RFCs/0011-partiql-insert-update-spec.md

```diff
@@ -0,0 +1,761 @@
+            <attr name> = { <value expr> | DEFAULT }
+            | ( <attr name> [, <attr name>]... ) = <value>
+            | ( <attr name> [, <attr name>]... ) = ( <sub-select> )
+        }
+
+<primary key> ::= <attr name>
+
+<composite primary key> ::= <attr name>, <attr name> [, <attr name> ]...
+
+<value expr> ::= <partiql value>
+     | <sql value expr>
+
+<partiql value> ::= NULL
+    | <ion value>
+    | <tuple value>
+    | <collection value>
+
+<sql value expr> ::= <sql expression>
+
+<ion value> ::= <backtick> <ion literal> <backtick>
+
+<tuple value> ::= <l curly brace> <r curly brace>
+    | <l curly brace> <string value> : <partiql value> [, <string value> : <partiql value>]... <r curly brace>
+
+<collection value> ::= <array value>
+    | <bag value>
+
+<array value> ::= <left bracket> { <partiql value> [, <partiql value>]... } <right bracket>
+
+<bag value> ::= <opening bag> { <partiql value> [, <partiql value>]... } <closing bag>
+
+<sub-select> ::= <sfw query>
+
+(* See sfw_query in PartiQL spec. figure 3: BNF Grammar for PartiQL Queries:
+https://partiql.org/assets/PartiQL-Specification.pdf *)
+<sfw query>
+
+<index attr name> ::= <attr name>
+<attr name> ::= <identifier>
+<constraint name> ::= <identifier>
+
+(* Literals *)
+<l curly brace> ::=
+“{”
+<r curly brace> ::=
+“}”
+<left bracket> ::=
+“[”
+<right bracket> ::=
+“]”
+<backtick> ::=
+`
+<opening bag> ::=
+<<
+<closing bag> ::=
+>>
+<left paren> ::=
+  “(”
+ <right paren> ::=
+  “)”
+```
+
+## 3. Description
+
+Implementation based on this specification MUST lead to atomic insertions of PartiQL values into a Schema (E.g. a database table). Atomic insertions means that either all of the insertions specified by the INSERT statement are performed, or none of the insertions are performed—E.g. when there is an error in insertion of one or more data items.
+
+### 3. 1 Attribute names
+
+As providing attribute names is optional, when attribute names are omitted and values are provided by `VALUES` clause or a `<sub-select>`:
+
+
+* (a) If the number of values in a data item is equal to the number of declared attributes, the values are assumed to refer to the declared attributes, in order of their definition in the table schema, and
+
+* (b) If the number of values in data item (M) is fewer than the number of declared attributes (N) the values are assumed to refer to the first M declared attributes, in order, and
+
+* (c) If the number of values in a data item (M) is greater than the number of declared attributes (N) it is an error, and
+* (d) If a value is not specified for every required attribute, it is an error.
+
+
+It’s worth emphasizing that, the implementing database MUST provide an ordering for the declared attributes that appear in the Schema.
+
+When attribute names are omitted and `VALUES <bag value>` is provided, the values in `<bag value>` MUST include required attribute–value pairs and MAY include optional or undeclared attribute-value pairs.
```

Comment:
There's more nuances to this, I think.  Consider

```sql
CREATE TABLE foo (
    id int not null primary key,
    is_deleted boolean not null default false,
    name varchar(50)
)
```

Then: 
```
INSERT INTO foo (1, 'bob') would not be a valid statement.
```

The statement becomes valid only if `is_deleted` is moved to the last position in the table.

Reply:
First off, I can see a typo here, it should be:
> When attribute names are omitted and `<bag value>` is provided ...

With that we have:
```SQL
INSERT INTO foo <<{id: 1, name: 50}>>;
```

Which intuitively should work, however based on the current spec. it must error out b/c of the following:
> ...the values in `<bag value>` MUST include required attribute–value pairs and MAY include optional or undeclared attribute-value pairs.

Having said that, I can see a nuance related to `default` value that we need to specify/call-out in this case.

Now if we want to insert using `VALUES`, it'll be:

```SQL
INSERT INTO foo VALUES (1, 'bob');
```

With this, seems like it falls into category `(b)` b/c number of values is fewer than number of _declared attributes_ (2 < 3). In that case, it'll be an `TypeMismatch` error or something like that—if the implementing NoSQL database provides the ordering from top to bottom for an Item. I don't see an issue with this though.

> (b) If the number of values in data item (M) is fewer than the number of declared attributes (N) the values are assumed to refer to the first M declared attributes, in order,

Reply:
Sorry I omitted `VALUES` from the original insert statement which was probably confusing.